### PR TITLE
Defer move generation to nodes that actually need it

### LINKED ISSUE #11: Search: defer per-node move generation — ~50% of generated move lists are never read (+38% NPS measured)
## Problem / opportunity

Every node in `PrincipalVariationSearch` and `QuiescenceSearch`
(`NeraChessSearch/src/SearchEngine.cpp`) begins with

```cpp
Score terminalScore;
if (IsDrawOrTerminal(board, terminalScore, ply))
    return terminalScore;
```

`IsDrawOrTerminal` calls `ChessBoard::GetGameOver(true)`
(`NeraChessEngine/src/ChessBoard.cpp:615`), which unconditionally runs
`m_MoveGenerator.GenerateMoves(...)` because checkmate/stalemate is detected as
"the legal move list is empty". `ChessBoard::IsInCheck()` (`ChessBoard.cpp:903`)
does the same thing for the same reason.

So a full legal move generation is paid **before** the TT probe, before reverse
futility, before the null-move cutoff, and before the quiescence stand-pat
cutoff — all of which can return without ever looking at a move.

There is a second instance of the same pattern inside quiescence: the delta and
SEE pruning decision is taken *after* `MakeSearchMove`:

```cpp
MakeSearchMove(board, move);
if (!inCheck && !board.IsInCheck() && (deltaPrunable || seePrunable))
{
    UndoSearchMove(board, move);
    continue;
}
```

A pruned capture therefore costs `DescribeMove` + `MakeMove` + an accumulator
push (two 1 KB `Simd::Copy`s plus the delta, `Architecture::HiddenSize == 512`,
`Weight == int16_t`) + a **full legal move generation of the child position**
(via `board.IsInCheck()`) + `UndoMove` + `Pop`. `ChessBoard::GivesCheck` already
exists for exactly this purpose and its own doc comment says so:

> Making the move and calling `IsInCheck()` answers the same question but forces
> a full legal-move generation for the child position, which is wasted work for
> a move the search is about to discard.

The main search was already converted to that pattern (see
`docs/IMPROVEMENT_BACKLOG.md`, "Pruning decisions moved ahead of `MakeMove`");
quiescence was not. `StaticExchangeEvaluation` is also called unconditionally
here even when delta pruning has already fired, and the same SEE value was
already computed for the same move a few lines earlier in `SortMoves`.

### Measured on the current tree

Instrumented counters, 10 positions at fixed `go depth 11`, book off,
1 thread, 64 MiB hash, `9e5d158`, Intel Xeon @ 2.80 GHz:

| Counter | Value |
| --- | ---: |
| PVS nodes | 501,465 |
| quiescence nodes | 638,349 (56.0% of nodes) |
| quiescence nodes returning on a TT cutoff or stand-pat, i.e. never reading the move list | 335,459 = **52.6%** |
| PVS nodes that actually reach the move loop | **48.2%** |
| quiescence capture candidates pruned *after* make/unmake | 230,348 of 554,627 = **41.5%** |

`callgrind` on a single `go depth 9` (book-load noise excluded) puts move
generation at roughly a third of search instructions —
`MoveGenerator::CalculateSlidingMoves` alone is 7.7% of all instructions
retired, ahead of everything except `Network::Forward`.

## Why this may improve NeraChess

This is pure throughput at unchanged search shape: the same tree is searched,
with fewer instructions per node. `docs/ENGINE_STRENGTH.md` measures the engine
at a fixed time control, so a node-rate gain converts directly into extra depth.
It also compounds with the selectivity work already shipped — the more nodes
that are cut before the move loop, the larger the fraction of move generation
that is wasted, so the current pruning layer is what makes this worth doing now.

## Proposed change or experiment

Two independent parts. Both were prototyped locally and measured separately.

**A. Take the quiescence pruning decision before `MakeSearchMove`.**
Use the existing `board.GivesCheck(move)` instead of make → `IsInCheck()` →
unmake, and short-circuit SEE behind the delta test:

```cpp
if (!inCheck && !(move.GetMoveFlags() & MoveFlags::IS_PROMOTION))
{
    const Piece victim = /* unchanged */;
    const bool prunable = standPat + PieceValue(victim) + 200 < alpha ||
        MoveOrdering::StaticExchangeEvaluation(board, move) < -50;
    if (prunable && !board.GivesCheck(move))
        continue;
}
MakeSearchMove(board, move);
```

This is behaviour-identical by construction. Reusing the SEE values `SortMoves`
already computed (carrying them out of the sort rather than recomputing) is a
further win and is the prerequisite the backlog notes for retrying SEE pruning
of quiet moves in the main search.

**B. Generate moves only when the node needs them.**
Needs two cheap `ChessBoard` helpers that touch no move generator state:

- an attack-table-based "is the side to move in check" test (same shape as the
  existing `GivesCheck`, aimed at one's own king), and
- a draw test covering only the conditions that do not need a move list
  (50-move, threefold, insufficient material).

Then reorder both node prologues to: cheap draw test → mate-distance bounds →
`depth <= 0` → TT probe → cheap in-check → static eval / reverse futility /
null move / stand-pat → **generate** → `if (moves.size() == 0) return inCheck ?
-SCORE_MATE + ply : SCORE_DRAW;` → sort. `MakeNullMove`'s own `IsInCheck()`
guard and the in-loop `givesCheck = board.IsInCheck()` should use the cheap test
too, otherwise they reintroduce the generation this removes. Exposing a
`const MoveList<218>&` accessor alongside `GetLegalMoves()` also drops an
~880-byte copy in the quiescence capture filter.

Design constraint worth stating explicitly: with (B), a **stalemate** node whose
stand-pat, reverse-futility or null-move test fires returns that score instead
of `SCORE_DRAW`, because no move list was generated to reveal that the position
is terminal. This is what every mainstream engine does (Stockfish resolves
`!moveCount` after the move loop for the same reason), and the measured node
delta was 0.009%. If it is judged unacceptable, the fallback is to apply (B) to
the TT-cutoff path only, which keeps roughly the smaller half of the win.

## Risks / trade-offs

- The cheap in-check test is new code on the hottest path in the engine; if it
  disagrees with the generator anywhere, checkmate detection and check evasions
  break silently. It must be verified against `IsInCheck()` exhaustively, not
  just spot-checked.
- (B) changes stalemate handling as described above. Positions where a side is
  stalemated *and* the static eval is above beta are rare but not impossible in
  endgames; a fortress/stalemate test position is worth adding to
  `NeraChessTests/src/main.cpp`.
- Neither part changes the tree, so neither is expected to show up as an Elo
  gain at fixed depth or fixed nodes — only at fixed time. Reporting a fixed-node
  match as "no change" would be a misreading, not a refutation.
- `GivesCheck` is not free (1.2% of instructions today, and (A) increases how
  often it is called). The measurement below already includes that cost.

## Validation

Correctness first:

- `./bin/Release/NeraChessTests/NeraChessTests` and
  `--eval-file NeraChessApp/Resources/NNUE/nera.nnue` must both pass, including
  perft and the make/undo and hash invariants.
- Exhaustive agreement of the new in-check test with `IsInCheck()` over a perft
  walk. Locally: 11,024,491 positions across 6 seed FENs (startpos, kiwipete,
  the standard perft 3/4/5 positions and a middlegame), **0 mismatches**.
- Fixed-depth node counts and best moves must be unchanged (part A) or unchanged
  to within the stalemate edge case (part B).

Performance benchmark: fixed `go depth N` over a position set, comparing nodes
(should match) and wall time. Measured here, 10 positions, book off, 1 thread,
64 MiB hash:

| Build | depth 11 nodes | nps | depth 12 time | depth 13 time |
| --- | ---: | ---: | ---: | ---: |
| `9e5d158` baseline | 1,453,884 | 484.8k | 5.99 s | 10.41 s |
| + A only | 1,453,884 | 544.9k (+12.4%) | — | — |
| + B only | 1,454,014 | 556.5k (+14.8%) | — | — |
| + A + B | 1,454,014 | 669.3k (**+38.1%**) | 4.46 s (−26%) | 7.78 s (−25%) |

Best moves were identical for all 10 positions at depth 12. `--bench` (perft)
and `--nnue-bench` should also be recorded, since (B) touches `ChessBoard`.

Playing strength: a fixed-time self-play match is the thing that decides it —
the existing `10+0.1` Fastchess setup in `docs/Strength-Testing.md` / the
`strength-test` workflow, or SPRT with bounds around [0, 5]. A fixed-node match
should also be run and is expected to come back at 0 Elo; that is the check that
the tree really is unchanged.

## Expected result

Keep the change if fixed-depth node counts are preserved, every regression test
passes, the in-check equivalence walk is clean, and the fixed-time match is
non-negative with a confidence interval excluding zero on the positive side. A
~35% node-rate gain is worth roughly half a ply to a ply of depth; the honest
statement is that the *speed* result is measured and the *Elo* result is not,
and it needs the match to claim one.

### PR DESCRIPTION
**Pull request by raphaelhounsiagaman**
## Summary

`PrincipalVariationSearch` and `QuiescenceSearch` generate the full legal move list unconditionally at the top of every node, before the TT probe, reverse futility, null move, and quiescence stand-pat cutoffs that frequently return without ever reading a move. This defers move generation until a node actually survives every cutoff that doesn't need it, following up on the "Part A" quiescence-pruning reorder already shipped in #22.

Closes #11.

## Problem

Every node paid for `ChessBoard::GetGameOver(true)` / `IsInCheck()` up front, both of which unconditionally call `MoveGenerator::GenerateMoves`. That cost was paid even for nodes that immediately return via a TT cutoff, reverse futility, a null-move cutoff, or quiescence stand-pat -- exactly the paths issue #11 measured as ~50% of quiescence nodes and roughly half of PVS nodes.

## Implementation

- `ChessBoard::IsInCheck()` is now an attack-table lookup (pawn/knight/king masks plus the existing magic-bitboard slider lookups) that answers whether the side to move's king is attacked, without touching the move generator. The old move-generation-based implementation is kept as `IsInCheckByMoveGeneration()`, used only as a reference for testing.
- A new `ChessBoard::IsRuleDraw()` reports the draw conditions that don't need a move list -- 50-move rule, threefold repetition, insufficient material -- leaving checkmate/stalemate to be resolved from the generated list's emptiness once a node actually reaches move generation.
- `PrincipalVariationSearch` and `QuiescenceSearch` no longer call `IsDrawOrTerminal` (still used, unchanged, for the one-time root check in `SearchWorker`). Their prologues now run: `IsRuleDraw()` -> cheap `IsInCheck()` -> mate-distance clamp -> TT probe -> static eval / reverse futility / null move / stand-pat -> **generate** -> resolve checkmate/stalemate from an empty list -> sort/search moves.
- `ChessBoard::GetLegalMovesRef()` returns the cached move list by reference; quiescence's capture/check filter uses it to avoid an ~880-byte copy of a list it only reads.

## Why this approach

The draw test has to fire regardless of check status. My first version gated it on `!inCheck` on the reasoning that checkmate needs an empty move list to detect and "in check" implied that path -- but a position can be in check, have legal replies, and still be a repetition or 50-move draw, which is the common case for a repeated line of checks. Gating on `!inCheck` silently stopped detecting that and inflated node counts by >30% on a repeated-check endgame in local testing before the fix; the corrected version checks `IsRuleDraw()` unconditionally and accepts a much narrower edge case instead (a mate delivered by a move that also lands the halfmove clock on exactly 100 is reported as a draw), matching the tolerance the issue itself sets for the analogous stalemate case.

## Validation

```
Release regression suite (no network): PASS
Release regression suite (--eval-file nera.nnue): PASS
Debug regression suite (--eval-file nera.nnue, accumulator asserts active): PASS
--bench (perft to depth 6 from startpos): 119,060,324 nodes, matches known count
```

Added to the regression suite:
- `TestFastInCheck`: exhaustive `IsInCheck()` vs `IsInCheckByMoveGeneration()` comparison over a small tree from 10 seed FENs.
- `IsRuleDraw()` checked against the existing `GetGameOver` fixtures in `TestTerminalPositions`/`TestRepetition` (50-move, repetition, insufficient material, and the negative cases).

A standalone deep walk (depth 5 from 13 seed FENs, including the exact endgame positions used in the fixed-depth benchmark below) checked **376,234,995 positions** with **zero disagreements** between the fast `IsInCheck()` and the move-generation reference.

Fixed-depth comparison (`go depth 13`, single thread, book off, 64 MiB hash, `nera.nnue`, 10 positions covering openings/middlegames/drawish endgames):

| | baseline (3f32041) | candidate (92ce030) |
|---|---:|---:|
| Total nodes | 3,637,117 | 3,637,292 (+0.005%) |
| Best moves | -- | identical on all 10 positions |
| NPS (repeated runs) | ~772k-816k | ~873k-940k (+13% to +16%) |

The tiny node delta is the accepted stalemate/checkmate-adjacent edge case described above; it is not a tree-shape change.

## Strength test

Screening test per `docs/Strength-Testing.md`, 1,000 games at `10+0.1`, paired openings with reversed colors, 4 shards:

```
Candidate: 92ce030f2cd24dff35aa11632f65c087e0c5c92c
Baseline:  3f32041e8fef134435e81fac7c477018a9993a2c
W/D/L: 316 / 450 / 234
Candidate score: 54.10%
Estimated Elo: +28.6
Approximate paired 95% interval: [+14.9, +42.3]
Verdict: Likely improvement
Workflow run: https://github.com/raphaelhounsiagaman/NeraChess/actions/runs/33018830280
```

The interval is entirely above zero, but this is a screening result over 1,000 games, not proof of the exact Elo gain -- a longer test would narrow it further if that's ever needed.

## Risks

- The accepted edge case above (a mate landing exactly on halfmove 100, or a stand-pat/RFP/null-move/TT cutoff returning a non-terminal score for a position that also happens to be stalemate) trades a small, quantified inaccuracy for not paying for a move list on every node. Both are extremely rare and the strength test result already reflects their net effect.
- `IsInCheck()` is now new code on the hottest path in the engine (TT probe gating, RFP, null move, LMR all read it). It is verified exhaustively above, but any future change to piece movement rules would need to keep it in sync with the move generator's own check logic.

---
_Generated by [Claude Code](https://claude.ai/code/session_01Etb3hFAQjfnkF5VbsyDX7E)_

### CODE REVIEW

File: NeraChessSearch/src/SearchEngine.cpp [hunk 1]
```diff
@@ -430,9 +430,16 @@ namespace NeraChessSearch
...
+        // can be in check, have legal replies, and still be a repetition or
+        // 50-move draw, and that is the common case for a repeated check. The only
+        // conflict with checkmate is a mating move landing exactly on halfmove
+        // 100, which this reports as a draw; that is deliberately accepted rather
+        // than paying for a move list on every node to rule it out.
+        if (board.IsRuleDraw())
+            return SCORE_DRAW;
+        const bool inCheck = board.IsInCheck();
         if (ply >= MAX_PLY - 1)
             return EvaluateNode(board);
```
[the remark this answers is not in the document]
Comment (raphaelhounsiagaman) [reply NeraChessSearch/src/SearchEngine.cpp:451]: Confirmed, this was a real regression from deferring move generation past `IsDrawOrTerminal`: `IsRuleDraw()` only covers the draw conditions that don't need a move list, so checkmate/stalemate at the `MAX_PLY - 1` horizon fell through to a static evaluation instead. Fixed in a5dd6c0 by generating the move list at that boundary and returning the mate/draw score when it's empty, in both `PrincipalVariationSearch` and `QuiescenceSearch`. This path is only reachable via an unusually long forcing sequence, so paying for move generation there is free in practice. Full regression suite passes in Release and Debug, and fixed-depth node counts are unchanged (the horizon isn't reached at normal search depths).

---
_Generated by [Claude Code](https://claude.ai/code)_
